include/ginkgo/core/base/math.hpp

```diff
@@ -1274,6 +1297,48 @@ nan()
 }
 
 
+/**
+ * Checks if a floating point number is a quiet NaN (gko::nan()).
+ *
+ * @tparam T  type of the value to check
+ *
+ * @param value  value to check
+ *
+ * @return `true` if the value is bitwise equal to gko::nan<T>().
+ */
+template <typename T>
+GKO_INLINE GKO_ATTRIBUTES std::enable_if_t<!is_complex_s<T>::value, bool>
+is_nan_exact(const T& value)
+{
+    using type = typename detail::float_to_bytes_impl<T>::type;
+    type value_bytes{};
+    type nan_bytes{};
+    auto nan_value = nan<T>();
+    using std::memcpy;
+    memcpy(&value_bytes, &value, sizeof(value));
+    memcpy(&nan_bytes, &nan_value, sizeof(value));
+    return value_bytes == nan_bytes;
+}
+
+
+/**
+ * Checks if any component of a complex value is a quiet NaN (gko::nan).
+ *
+ * @tparam T  complex type of the value to check
+ *
+ * @param value  complex value to check
+ *
+ * @return `true` if any component of the complex number fulfills
+ * is_nan_exact(component).
+ */
+template <typename T>
+GKO_INLINE GKO_ATTRIBUTES std::enable_if_t<is_complex_s<T>::value, bool>
+is_nan_exact(const T& value)
+{
+    return is_nan_exact(value.real()) || is_nan_exact(value.imag());
+}
```

Comment:
Can you move these function to `common_trs_kernels`?
To me, they are not general enough to justify putting them in the public header. The chances that they are misused are quite big (this function likely doesn't detect NaNs caused by actual computations).

Reply:
makes sense, I'll move it